### PR TITLE
Make speed tool compatible with OpenSSL

### PR DESCRIPTION
**Pull request by billbo-yang**
### Description of changes: 
Currently, AWS-LC's bssl speed tool is only compatible with AWS-LC. This PR makes the tool compatible with both BoringSSL and OpenSSL in order to facilitate easier performance comparisons between the AWS-LC and the two libraries.

Instead of running `bssl speed ...` in order to benchmark performance, AWS-LC will now generate three tools when provided with their corresponding compiler flags. These tools take the same arguments as the BoringSSL Speed tool does.

Additionally, the speed tool now prints a message when it is benchmarking a non-release build of AWS-LC.

|  Tool Name  |  Compiler Flag  |
| ------------- | ------------- |
|  `awslc_bm`  |  `-DAWSLC_INSTALL_DIR`  |
|  `bssl_bm`  |  `-DBORINGSSL_INSTALL_DIR`  |
|  `ossl_bm`  |  `-DOPENSSL_INSTALL_DIR`  |

### Call-outs:
No significant architecture or design changes.

One thing to note: currently the tool expects for the libraries to be installed in a folder named `build` that is inside the base directory of each provided install directory. For example, if BoringSSL was cloned to `/usr/boringssl/`, AWS-LC will look in `/usr/boringssl/build/` when building in order to find necessary files to create the BoringSSL compatible benchmarking tool.

### Testing:
This change was tested by building the program with the compiler flags provided, running all tests on each benchmark, then comparing the output of to see expected performance differences were there.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tool/speed.cc [hunk 1]
```diff
@@ -1320,70 +1331,66 @@ bool Speed(const std::vector<std::string> &args) {
...
+     !SpeedHash(EVP_sha256(), "SHA-256", selected) ||
+     !SpeedHash(EVP_sha384(), "SHA-384", selected) ||
+     !SpeedHash(EVP_sha512(), "SHA-512", selected) ||
+     !SpeedRandom(selected) ||
+     !SpeedECDH(selected) ||
+     !SpeedECDSA(selected) ||
+     !SpeedScrypt(selected) ||
+     !SpeedRSA(selected) ||
+     !SpeedRSAKeyGen(selected)
+#if !defined(OPENSSL_BENCHMARK)
```
Comment (torben-hansen) [inline tool/speed.cc:1365]: We miss benchmarks for a number of important algorithms:
* Curve25519-based
* AES-CBC
* AES-GCM

Why not bridge that gap?

File: tool/ossl_bm.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
+//#include <openssl/aead.h>
```
Comment (torben-hansen) [inline tool/ossl_bm.h:1]: Oopsie? :)

File: tool/speed.cc (hunk 1 above)
Comment (billbo-yang) [reply tool/speed.cc:1365]: Unless I'm not understanding something, I think the benchmarks you want are at lines 1367-1368, 1371-1372, and 1380. Or is it that these aren't the benchmarks you're looking for...

Comment (torben-hansen) [reply tool/speed.cc:1365]: That's right. Also chacha20-poly1305.

It's not completely straightforward though, because the APIs differ. So, might need some extra benchmark functions that does it in a compatible, unified, and fair way.

File: tool/ossl_bm.h (hunk 2 above)
Comment (billbo-yang) [reply tool/ossl_bm.h:1]: Oops I'll get rid of those

File: tool/CMakeLists.txt [hunk 3]
```diff
@@ -29,17 +31,62 @@ endif()
 
 if(APPLE OR WIN32 OR ANDROID)
   target_link_libraries(bssl ssl crypto)
+  set(LIBRT_FLAGS "")
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:34]: Nit: `LIBRT_FLAG` I guess, since there is only one?

File: tool/CMakeLists.txt [hunk 4]
```diff
@@ -29,17 +31,62 @@ endif()
...
+          args.cc
+          const.cc
+          benchmark.cc
+  )
+  # Link with the internal tool directory for shared headers with the rest of the tool instead of the top level AWS-LC
+  # include directory
+  target_include_directories(${target_name} PUBLIC ${additional_include_dir} ${AWSLC_INSTALL_DIR}/include/internal/tool)
+  target_link_libraries(${target_name} ${lib_crypto} ${LIBRT_FLAGS})
+
+  string(TOUPPER ${target_name} upper_target_name)
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:60]: why this?

File: tool/CMakeLists.txt [hunk 5]
```diff
@@ -29,17 +31,62 @@ endif()
...
+  target_compile_options(${target_name} PUBLIC -D${upper_target_name})
+endfunction()
+
+if(AWSLC_INSTALL_DIR)
+  build_benchmark(awslc_bm ${AWSLC_INSTALL_DIR}/include crypto)
+endif()
+
+# This expects a directory which contains lib/libcrypto.a and the includes in include/openssl/*.h
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:68]: And that OpenSSL artifacts are under `build` as well.

File: tool/CMakeLists.txt [hunk 6]
```diff
@@ -29,17 +31,62 @@ endif()
...
+if(OPENSSL_INSTALL_DIR)
+  build_benchmark(ossl_bm ${OPENSSL_INSTALL_DIR}/include ${OPENSSL_INSTALL_DIR}/build/lib/libcrypto.a)
+
+  # libpthread is a linux only library
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:72]: > linux only library

so, what's BSD then? :)

File: tool/CMakeLists.txt [hunk 7]
```diff
@@ -29,17 +31,62 @@ endif()
...
+  if(NOT MSVC AND NOT ANDROID)
+    # OpenSSL uses pthread_rwlock_unlock from pthread and  dlopen/close from dl
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:74]: Quite a bit more symbols than that.

File: tool/CMakeLists.txt [hunk 8]
```diff
@@ -29,17 +31,62 @@ endif()
...
+    target_compile_options(ossl_bm PUBLIC -DOPENSSL_BENCHMARK -Wno-gnu-anonymous-struct -Wno-nested-anon-types)
+  endif()
+endif()
+
+if(BORINGSSL_INSTALL_DIR)
+  build_benchmark(bssl_bm ${BORINGSSL_INSTALL_DIR}/include ${BORINGSSL_INSTALL_DIR}/build/crypto/libcrypto.a)
+
+  # libpthread is a linux only library
+  if(NOT MSVC AND NOT ANDROID)
+    # BoringSSL uses pthread_rwlock_unlock from pthread
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:86]: Quite a bit more symbols than that.

File: tool/CMakeLists.txt [hunk 9]
```diff
@@ -29,17 +31,62 @@ endif()
...
+    target_link_libraries(bssl_bm pthread)
+    # BoringSSL uses both anonymous structs and unions
+    target_compile_options(bssl_bm PUBLIC -Wno-gnu-anonymous-struct -Wno-nested-anon-types)
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:89]: The speed tool uses anom structs and unions? Not sure I understand why these are needed.

File: tool/args.cc [hunk 10]
```diff
@@ -71,16 +70,20 @@ bool ParseKeyValueArguments(std::map<std::string, std::string> *out_args,
...
   for (size_t i = 0; templates[i].name[0] != 0; i++) {
-    const struct argument *templ = &templates[i];
+    const argument_t *templ = &templates[i];
     fprintf(stderr, "%s\t%s\n", templ->name, templ->description);
   }
 }
 
 bool GetUnsigned(unsigned *out, const std::string &arg_name,
                  unsigned default_value,
+#if !defined(OPENSSL_BENCHMARK)
```
Comment (torben-hansen) [inline tool/args.cc:82]: Why is this needed when: `typedef std::map<std::string, std::string> args_map_t;`?

File: tool/benchmark.cc [hunk 11]
```diff
@@ -0,0 +1,12 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "internal.h"
+
+int main(int argc, char **argv) {
+  args_list_t args;
+  for (int i = 1; i < argc; i++) {
```
Comment (torben-hansen) [inline tool/benchmark.cc:8]: `int` -> `size_t`

File: tool/bssl_bm.h [hunk 12]
```diff
@@ -0,0 +1,23 @@
+#pragma once
```
Comment (torben-hansen) [inline tool/bssl_bm.h:1]: Rest of library uses standard header guards.

File: tool/const.cc [hunk 13]
```diff
@@ -17,7 +17,6 @@
 
 #include "internal.h"
 
-
```
Comment (torben-hansen) [inline tool/const.cc:20]: Good fix :)

File: tool/speed.cc [hunk 14]
```diff
@@ -24,23 +24,15 @@
...
-#include <openssl/mem.h>
-#include <openssl/nid.h>
-#include <openssl/rand.h>
-#include <openssl/rsa.h>
-#include <openssl/trust_token.h>
+#include "internal.h"
+
+#if !defined(OPENSSL_BENCHMARK)
+#include "bssl_bm.h"
+#include <../crypto/ec_extra/internal.h>
```
Comment (torben-hansen) [inline tool/speed.cc:31]:
```suggestion
#include "../crypto/ec_extra/internal.h"
```

File: tool/speed.cc [hunk 15]
```diff
@@ -202,11 +196,20 @@ static bool SpeedRSA(const std::string &selected) {
...
   for (unsigned i = 0; i < OPENSSL_ARRAY_SIZE(kRSAKeys); i++) {
+#else
+  for(unsigned i = 0; i < BM_ARRAY_SIZE(kRSAKeys); i++) {
+#endif
     const std::string name = kRSAKeys[i].name;
 
-    bssl::UniquePtr<RSA> key(
-        RSA_private_key_from_bytes(kRSAKeys[i].key, kRSAKeys[i].key_len));
+//    BM_NAMESPACE::UniquePtr<RSA> key(RSA_private_key_from_bytes(kRSAKeys[i].key, kRSAKeys[i].key_len));
+    /* d2i_RSAPrivateKey expects to be able to modify the input pointer as it parses the input data and we don't want it
```
Comment (torben-hansen) [inline tool/speed.cc:207]: Use `//`.

File: tool/speed.cc [hunk 16]
```diff
@@ -202,11 +196,20 @@ static bool SpeedRSA(const std::string &selected) {
     {"RSA 4096", kDERRSAPrivate4096, kDERRSAPrivate4096Len},
   };
 
+#if !defined(OPENSSL_BENCHMARK)
```
Comment (torben-hansen) [inline tool/speed.cc:199]: Can't we just use `BM_ARRAY_SIZE` for everything?

File: tool/speed.cc [hunk 17]
```diff
@@ -202,11 +196,20 @@ static bool SpeedRSA(const std::string &selected) {
...
+#else
+  for(unsigned i = 0; i < BM_ARRAY_SIZE(kRSAKeys); i++) {
+#endif
     const std::string name = kRSAKeys[i].name;
 
-    bssl::UniquePtr<RSA> key(
-        RSA_private_key_from_bytes(kRSAKeys[i].key, kRSAKeys[i].key_len));
+//    BM_NAMESPACE::UniquePtr<RSA> key(RSA_private_key_from_bytes(kRSAKeys[i].key, kRSAKeys[i].key_len));
+    /* d2i_RSAPrivateKey expects to be able to modify the input pointer as it parses the input data and we don't want it
+     * to modify the original *key data. Therefore create a new temp variable that points to the same data and pass
```
Comment (torben-hansen) [inline tool/speed.cc:208]:
```suggestion
     * to modify the original |*key| data. Therefore create a new temp variable that points to the same data and pass
```

File: tool/speed.cc [hunk 18]
```diff
@@ -202,11 +196,20 @@ static bool SpeedRSA(const std::string &selected) {
...
+     * in the reference to it. As a sanity check make sure input_key points to the end of the *key.
```
Comment (torben-hansen) [inline tool/speed.cc:209]:
```suggestion
     * in the reference to it. As a sanity check make sure |input_key| points to the end of the |*key|.
```

File: tool/speed.cc [hunk 19]
```diff
@@ -202,11 +196,20 @@ static bool SpeedRSA(const std::string &selected) {
...
+     */
+    const uint8_t *input_key = kRSAKeys[i].key;
+    BM_NAMESPACE::UniquePtr<RSA> key(d2i_RSAPrivateKey(NULL, &input_key, (long) kRSAKeys[i].key_len));
```
Comment (torben-hansen) [inline tool/speed.cc:212]: Since the signature is `TYPE *d2i_TYPE(TYPE **a, const unsigned char **ppin, long length);`, you are relying on that `uint8_t` is a typedef for `unsigned char`. Which it typically is.

File: tool/speed.cc (hunk 1 above)
Comment (torben-hansen) [reply tool/speed.cc:1365]: Let's move ahead without this.

File: .gitignore [hunk 20]
```diff
@@ -30,3 +30,6 @@ util/bot/win_toolchain.json
 util/bot/yasm-win32.exe
 
 test_build_dir
+
+files
```
[the remark this answers is not in the document]
Comment (billbo-yang) [reply .gitignore:34]: It's a folder that I was using to hold some benchmarking data while doing some testing - don't know how it was added to the .gitIgnore :/

Will remove :)

**Review by billbo-yang — COMMENTED** [review]
Looks like I missed quite a few things when cleaning up my implementation. Will go back through and double/triple check everything.

File: tool/CMakeLists.txt [hunk 21]
```diff
@@ -29,17 +31,62 @@ endif()
...
+    set(LIBRT_FLAGS "")
   endif()
 endif()
 
-install(TARGETS bssl
-        EXPORT bssl-targets
-        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
-)
+function(build_benchmark target_name additional_include_dir lib_crypto)
+  message(${additional_include_dir})
```
[the remark this answers is not in the document]
Comment (billbo-yang) [reply tool/CMakeLists.txt:47]: I intended it to act as a sort of confirmation of what paths were provided so that the user could double check whether the correct path was provided.

File: tool/CMakeLists.txt (hunk 6 above)
Comment (billbo-yang) [reply tool/CMakeLists.txt:72]: What does BSD stand for?

File: tool/CMakeLists.txt (hunk 7 above)
Comment (billbo-yang) [reply tool/CMakeLists.txt:74]: Do you know any references I can look at to find the others?

File: tool/benchmark.cc (hunk 11 above)
Comment (billbo-yang) [reply tool/benchmark.cc:8]: Any particular reason for this?

File: tool/speed.cc (hunk 14 above)
Comment (billbo-yang) [reply tool/speed.cc:31]: Implementing this change gives us following error message during build:
`FAILED: tool/CMakeFiles/bssl_bm.dir/speed.cc.o 
/usr/bin/c++  -DBORINGSSL_DISPATCH_TEST -DBORINGSSL_HAVE_LIBUNWIND -DBORINGSSL_IMPLEMENTATION -I../third_party/googletest/include -I/home/ubuntu/dev/boringssl/include -I/home/ubuntu/dev/awslc/include/internal/tool -DAWS_LC_STDALIGN_AVAILABLE -DAWS_LC_BUILTIN_SWAP_SUPPORTED -Wvla -Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -ggdb -Wall -fvisibility=hidden -fno-common -Wno-free-nonheap-object -Wimplicit-fallthrough -Wmissing-declarations -std=c++11 -fno-exceptions -fno-rtti -Wshadow -fPIE -MD -MT tool/CMakeFiles/bssl_bm.dir/speed.cc.o -MF tool/CMakeFiles/bssl_bm.dir/speed.cc.o.d -o tool/CMakeFiles/bssl_bm.dir/speed.cc.o -c ../tool/speed.cc
In file included from /home/ubuntu/dev/boringssl/include/openssl/stack.h:62:0,
                 from /home/ubuntu/dev/boringssl/include/openssl/ex_data.h:114,
                 from /home/ubuntu/dev/boringssl/include/openssl/ec_key.h:75,
                 from /home/ubuntu/dev/boringssl/include/openssl/ec.h:391,
                 from ../tool/bssl_bm.h:11,
                 from ../tool/speed.cc:30:
../tool/../crypto/ec_extra/../fipsmodule/ec/internal.h:95:23: error: expected string-literal before ‘bn__small_functions_not_usable’
                       bn__small_functions_not_usable)
                       ^
/home/ubuntu/dev/boringssl/include/openssl/type_check.h:69:62: note: in definition of macro ‘OPENSSL_STATIC_ASSERT’
 #define OPENSSL_STATIC_ASSERT(cond, msg) static_assert(cond, msg)
                                                              ^~~
../tool/../crypto/ec_extra/../fipsmodule/ec/internal.h:95:23: error: expected ‘)’ before ‘bn__small_functions_not_usable’
                       bn__small_functions_not_usable)
                       ^
/home/ubuntu/dev/boringssl/include/openssl/type_check.h:69:62: note: in definition of macro ‘OPENSSL_STATIC_ASSERT’
 #define OPENSSL_STATIC_ASSERT(cond, msg) static_assert(cond, msg)
                                                              ^~~
In file included from ../tool/../crypto/ec_extra/internal.h:20:0,
                 from ../tool/speed.cc:31:
../tool/../crypto/ec_extra/../fipsmodule/ec/internal.h:103:1: error: expected ‘;’ before ‘typedef’
 typedef union {
 ^~~~~~~
`

File: tool/CMakeLists.txt (hunk 21 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply tool/CMakeLists.txt:46]: I think the message would need more context to be meaningful to a user. And why only the include path?

File: tool/CMakeLists.txt (hunk 7 above)
Comment (torben-hansen) [reply tool/CMakeLists.txt:69]: I think we just need to link against pthread and the linker lib. No need to specify the actual functions used.

Also, if you put a list with function names here, it would definitely become out-of-date at some point.

File: tool/benchmark.cc (hunk 11 above)
Comment (torben-hansen) [reply tool/benchmark.cc:8]: Yeah:
* It's a nit.
* The type `size_t` is guaranteed to be big enough to be able to index into the maximal possible array size.
* `int` is unsigned. We don't want to go to negative indexes.

In addition (not applicable above), sometimes you want to do range checks using the index variable. If the variable is unsigned, you need to think more.

File: tool/speed.cc (hunk 14 above)
Comment (torben-hansen) [reply tool/speed.cc:31]: Hmm something is up with the static assert macro `OPENSSL_STATIC_ASSERT`.

Comment (torben-hansen) [reply tool/speed.cc:31]: Okay I see what is going on. You are including the definition of `OPENSSL_STATIC_ASSERT` from BoringSSL (through `bssl_bm.h`). But you are also including internal header files from AWS-LC that expects the AWS-LC version of `OPENSSL_STATIC_ASSERT`. The definition of `OPENSSL_STATIC_ASSERT` in AWS-LC is not compatible with the BoringSSL definition, and hence you get the above errors.

One clue is that you see the following line in the error output:
```
../tool/../crypto/ec_extra/../fipsmodule/ec/internal.h:103:1: error: expected ‘;’ before ‘typedef’ typedef union { ^~~~~~~
```
This can't be the AWS-LC version because it doesn't use a union. It uses a struct: https://github.com/awslabs/aws-lc/blob/main/include/openssl/type_check.h#L90.

Where do you expect this the header file to come from?

You will have the same problem with the line below:
```
#include "../crypto/trust_token/internal.h"
```
if it starts defining things that is not compatible between AWS-LC and BoringSSL.

File: tool/benchmark.cc (hunk 11 above)
Comment (billbo-yang) [reply tool/benchmark.cc:8]: But in this case, argc is an int and afaik it is impossible for it to be negative. I'd have to cast it to a size_t to make it work if I made `int i` type size_t, which I think just introduces clutter. Obv this isn't the biggest issue in the world but I still don't really get using size_t here...

File: tool/CMakeLists.txt (hunk 21 above)
[the remark this answers is not in the document]
Comment (billbo-yang) [reply tool/CMakeLists.txt:46]: True, I'll go through and put some more context behind it

File: tool/speed.cc (hunk 14 above)
Comment (billbo-yang) [reply tool/speed.cc:31]: So instead of changing `<../crypto/ec_extra/internal.h>` to `"../crypto/ec_extra/internal.h"` wouldn't it make more sense for me to change `"../crypto/trust_token/internal.h"` to `"<../crypto/trust_token/internal.h>`? Or is there some convention against that?

File: tool/speed.cc (hunk 19 above)
Comment (billbo-yang) [reply tool/speed.cc:212]: Unless I'm missing something, `d2i_RSAPrivateKey` takes a `uint8_t` though? Or am I looking at the wrong thing

Comment (torben-hansen) [reply tool/speed.cc:204]: I see, AWS-LC and OpenSSL have different function signatures. Great.

I think this is fine because `uint8_t` is often the same as `unsigned char`.

File: tool/CMakeLists.txt [hunk 22]
```diff
@@ -29,17 +30,57 @@ endif()
...
+    target_compile_options(awslc_bm PUBLIC -DCMAKE_BUILD_TYPE)
+  endif()
+endif()
+
+# This expects a directory which contains the includes are in include/openssl/*.h and the OpenSSL artifacts in build/
+if(OPENSSL_INSTALL_DIR)
+  build_benchmark(ossl_bm ${OPENSSL_INSTALL_DIR}/include ${OPENSSL_INSTALL_DIR}/build/lib/libcrypto.a)
+
+  if(NOT MSVC AND NOT ANDROID)
+    target_link_libraries(ossl_bm -DOPENSSL_BENCHMARK pthread dl)
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:73]: Do you need this flag when linking?

File: tool/CMakeLists.txt [hunk 23]
```diff
@@ -29,17 +30,57 @@ endif()
...
+    target_compile_options(ossl_bm PUBLIC -DOPENSSL_BENCHMARK)
```
Comment (torben-hansen) [inline tool/CMakeLists.txt:74]: Don't you want this flag to be defined when you build the OpenSSL linked benchmark binary? What am I missing?

Comment (billbo-yang) [reply tool/CMakeLists.txt:74]: `OPENSSL_BENCHMARK` is added to the compile options for `ossl_bm` in this line in order to let the tool know that we want the OpenSSL linked benchmark binary to run some specific code for OpenSSL (i.e. include the right header file). If you look at lines 74 and 83 you can see I don't define this flag for the compilers for the awslc benchmark and BoringSSL benchmark

Comment (torben-hansen) [reply tool/CMakeLists.txt:74]: Yeah, but why not define the flag for Win and Android? Does it matter?

File: tool/README.md [hunk 24]
```diff
@@ -0,0 +1,44 @@
+# Tools for AWS-LC
+AWS-LC features enhanced benchmarking tools compatible with OpenSSL and BoringSSL in order to help facilitate 1-1 performance comparisons.
+
+## Benchmarking Tools
+When compiled, AWS-LC will generate three benchmarking tools when provided with corresponding compiler flags. These tools take the same arguments as the BoringSSL speed tool does.
+
+Additionally, the speed tool now prints a message when it is benchmarking a non-release build of AWS-LC.
```
Comment (torben-hansen) [inline tool/README.md:7]: Compared to what?

File: tool/README.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
+# Tools for AWS-LC
```
Comment (torben-hansen) [inline tool/README.md:1]: How do I use these tools?

File: tool/speed.cc [hunk 26]
```diff
@@ -1320,70 +1323,70 @@ bool Speed(const std::vector<std::string> &args) {
...
   static const size_t kTLSADLen = 13;
+
   // kLegacyADLen is the number of bytes that TLS passes to the "legacy" AEADs.
   // These are AEADs that weren't originally defined as AEADs, but which we use
   // via the AEAD interface. In order for that to work, they have some TLS
   // knowledge in them and construct a couple of the AD bytes internally.
   static const size_t kLegacyADLen = kTLSADLen - 2;
+#endif
+
+#if defined(CMAKE_BUILD_TYPE)
```
Comment (torben-hansen) [inline tool/speed.cc:1338]: Is this "hard" to overlook?

File: tool/CMakeLists.txt [hunk 27]
```diff
@@ -1,4 +1,3 @@
-include_directories(../include)
```
[the remark this answers is not in the document]
Comment (billbo-yang) [reply tool/CMakeLists.txt:1]: What do you mean by this?

File: tool/CMakeLists.txt [hunk 28]
```diff
@@ -29,17 +30,57 @@ endif()
...
+  # Link with the internal tool directory for shared headers with the rest of the tool instead of the top level AWS-LC
+  # include directory
+  target_include_directories(${target_name} PUBLIC ${additional_include_dir} ${AWSLC_INSTALL_DIR}/include/internal/tool)
+  target_link_libraries(${target_name} ${lib_crypto} ${LIBRT_FLAG})
+endfunction()
+
+if(AWSLC_INSTALL_DIR)
+  build_benchmark(awslc_bm ${AWSLC_INSTALL_DIR}/include crypto)
+
+  if(NOT CMAKE_BUILD_TYPE)
```
[the remark this answers is not in the document]
Comment (billbo-yang) [reply tool/CMakeLists.txt:63]: `NOT` is supposed to be there, if the `CMAKE_BUILD_TYPE` flag isn't present then we know we want to benchmark in debug mode.

File: tool/CMakeLists.txt (hunk 27 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply tool/CMakeLists.txt:1]: Can you somehow have the OpenSSL benchmark binary statically linked with OpenSSL, while the AWS-Lc binary is dynamically linked with AWS-LC?

Wanna prevent that if possible.

[the remark this answers is not in the document]
Comment (torben-hansen) [reply tool/CMakeLists.txt:1]: Let's punt to future work. Not terribly important imo.

**Review by torben-hansen — APPROVED** [review]

**Review by andrew-kaufman — APPROVED** [review]
